arbor/include/arbor/domain_decomposition.hpp

```diff
@@ -31,26 +32,37 @@ struct group_description {
 /// distribution of cells across cell_groups and domains.
 /// A load balancing algorithm is responsible for generating the
 /// domain_decomposition, e.g. arb::partitioned_load_balancer().
-struct domain_decomposition {
+class domain_decomposition {
+public:
+    domain_decomposition() = delete;
+    domain_decomposition(const recipe& rec, const context& ctx, const std::vector<group_description>& groups);
+
+    int gid_domain(cell_gid_type gid) const;
+    int num_domains() const;
+    int domain_id() const;
+    cell_size_type num_local_cells() const;
+    cell_size_type num_global_cells() const;
+    const std::vector<group_description>& groups() const;
```

Comment:
While reading the code I saw these two things used frequently
```cxx
auto g  = dd.groups()[ix];
auto n = dd.groups().size();
```
Maybe add this as final polish
```cxx
class domain_decomposition {
public:
  size_t num_groups();
  group_description& group(size_t);
  const group_description& group(size_t) const;
```

Reply:
I'll only add `const group_description& group(size_t) const;`. We don't want the groups to be mutable after construction.